control/state.py

```diff
@@ -229,53 +279,77 @@ def get_state(self) -> Dict[str, str]:
             omap_dict = dict(i)
         return omap_dict
 
-    def _add_key(self, key: str, val: str):
+    def _add_key(self, key: str, val: str, status: RequestStatus):
         """Adds key and value to the OMAP."""
-        try:
-            version_update = self.version + 1
-            with rados.WriteOpCtx() as write_op:
-                # Compare operation failure will cause write failure
-                write_op.omap_cmp(self.OMAP_VERSION_KEY, str(self.version),
-                                  rados.LIBRADOS_CMPXATTR_OP_EQ)
-                self.ioctx.set_omap(write_op, (key,), (val,))
-                self.ioctx.set_omap(write_op, (self.OMAP_VERSION_KEY,),
-                                    (str(version_update),))
-                self.ioctx.operate_write_op(write_op, self.omap_name)
-            self.version = version_update
-            self.logger.debug(f"omap_key generated: {key}")
-        except Exception as ex:
-            self.logger.error(f"Unable to add key to omap: {ex}. Exiting!")
+        version_update = self.version + 1
+        for i in range(self.state_update_retries):
+            try:
+                with rados.WriteOpCtx() as write_op:
+                    # Compare operation failure will cause write failure
+                    write_op.omap_cmp(self.OMAP_VERSION_KEY, str(self.version),
+                                      rados.LIBRADOS_CMPXATTR_OP_EQ)
+                    self.ioctx.set_omap(write_op, (key,), (val,))
+                    self.ioctx.set_omap(write_op, (self.OMAP_VERSION_KEY,),
+                                        (str(version_update),))
+                    self.ioctx.operate_write_op(write_op, self.omap_name)
+            except Exception as ex:  # Update and retry on failure
+                self.update_trigger.set()
```

Comment:
I don't think this will work. We need to retry the whole operation in case of version mismatch. An example where this will fail:

1. GW A adds `init_add_A_X`
2. GW B adds `init_add_B_X` (might fail because of version and needs to update but since key is init nothing happens and it can successfully add the init key after updating to the latest version)
3. Both reflect the change in SPDK and both succeed
4. Both try to add `X` (state == success)
5. B succeeds first and adds `X` to OMAP
6. A tries to add key and fails because version changed => gets new version and updates `X` with the value of B (calls SPDK etc)
7. The loop continues and overwrites `X` in the OMAP with the value of A => other nodes get notified and update themselves to latest version and set the value of `X` to A's value but A doesn't update itself because it made the change and thought it succeeded.

This is just one example where this fails. I think you can come up with more complex scenarios where this doesn't work.

Reply:
Or would this be covered by what you meant by ```At this time, the PR doesn't include logic to wait for conflicting in-progress requests to complete. If there aren't many changes requested, we can add it on.```

Reply:
I think either way I can come up with scenarios where the key is not directly conflicting but this still creates invalid state after retry (e.g. add_host on a subsystem that does not exist anymore).

Reply:
I see the problem now, but I'm not sure what you mean that `we need to retry the whole operation`. Unless you mean that we undo the SPDK call and then update and then re-issue the SPDK call?

Reply:
I think the correct order should be:

1. On conflict before update to latest version: undo SPDK operation
2. update to latest version
3. try to redo SPDK operation

Reply:
If I remember correctly we introduced this multi step approach because SPDK could fail on whatever the user provided. So what if we allow the RPC to be invalid and SPDK to fail on all nodes but instead of shutting down we just ignore the config option (put a warning in the log) and the GW that handles the RPC reverts the change after it fails locally? 
This also helps the case when we somehow create an invalid config that we can't fix because we fail the daemon if it can't be applied.

Reply:
In this case, it seems like the temporary keys are not needed, correct?

Reply:
Yes, I think so. I have to think a bit about it maybe there is a case where we need temporary keys.